### PR TITLE
fix(frontends/lean/builtin_exprs): "fix" "`a` bug"

### LINKED ISSUE #437: The `a` bug
### Prerequisites

* [X] Put an X between the brackets on this line if you have done all of the following:
    * Checked that your issue isn't already [filed](https://github.com/leanprover-community/lean/issues).
        * Specifically, check out the [wishlist](https://github.com/leanprover-community/lean/issues?q=is%3Aissue+is%3Aopen+label%3AI-wishlist), open [RFCs](https://github.com/leanprover-community/lean/issues?q=is%3Aissue+is%3Aopen+label%3ARFC),
        or [feature requests](https://github.com/leanprover-community/lean/issues?q=is%3Aissue+is%3Aopen+label%3AFeature).
    * Reduced the issue to a self-contained, reproducible test case.

### Description

The famous `a` bug doesn't need a description, we have all met it.

### Steps to Reproduce

```lean
example (a : ℕ)  : true :=
begin
  have : ∀ n, n ≥ 0 →  a ≤ a,
  sorry
end
```
**Expected behavior:** [What you expect to happen]

No error.

**Actual behavior:** [What actually happens]

Lean chooses the name `a` for the premise `n ≥ 0` and confuses itself.

**Reproduces how often:** [What percentage of the time does it reproduce?]

100%

### Versions

Lean (version 3.18.4, commit f539be1e867c, Release)

### PR DESCRIPTION
**Pull request by rwbarton**
In tactic mode, `p -> a` would be interpreted as `Π (a : p), a`,
capturing the variable `a`.
Change the binder name used by `->` to `_x` to avoid this.

Fixes #437.
(more or less)

### CODE REVIEW

**Comment by bryangingechen** [comment]
bors try

**Comment by gebner** [comment]
There is a lot of stuff that depends on the name `a`. :disappointed:  I wonder how much of mathlib will break.

File: tests/lean/run/437.lean [hunk 1]
```diff
@@ -0,0 +1,6 @@
+example (a : ℕ) : true :=
```
Comment (bryangingechen) [inline tests/lean/run/437.lean:1]: I think we've been trying to name test files that reference `leanprover-community` issue numbers with a `c`, so this should be called `437c.lean`.

**Comment by digama0** [comment]
@gebner It's not `a` bug it's `_x` feature. Depending on auto names is bad practice.

File: tests/lean/run/437.lean (hunk 1 above)
Comment (rwbarton) [reply tests/lean/run/437.lean:1]: Ah thanks, I wondered whether there was a convention for that, but I saw some other test files like `####b.lean`, so I didn't work out that `c` meant community.

**Comment by rwbarton** [comment]
Fixed some additional tests. Now trying to build mathlib to see how bad the damage is.

**Comment by rwbarton** [comment]
Hmm, now lean is segfaulting on `set_theory.pgame`, so maybe don't merge this just yet, though I don't know the cause yet.

**Comment by rwbarton** [comment]
The change in default binder name to `_x` is triggering #447, because `pgame` is defined as
```lean
inductive pgame : Type (u+1)
| mk : ∀ α β : Type u, (α → pgame) → (β → pgame) → pgame
```

I'm not sure how to proceed. I suppose it would be nice to use the name `_x` only in tactic mode and `a` elsewhere, as this would also cause less breakage from use of autogenerated names. But it doesn't seem to be that straightforward based on how `→` is parsed currently. Also possible would be to try to come up with a name not starting with `_` but which won't conflict with anything someone is likely to type; suggestions?

File: src/frontends/lean/builtin_exprs.cpp [hunk 2]
```diff
@@ -1090,9 +1090,11 @@ static expr parse_field(parser & p, unsigned, expr const * args, pos_info const
     }
 }
 
+static name * g_arrow_binder = nullptr;
+
 parse_table init_led_table() {
     parse_table r(false);
-    r = r.add({transition("->", mk_expr_action(get_arrow_prec()-1))},    mk_arrow(Var(1), Var(1)));
+    r = r.add({transition("->", mk_expr_action(get_arrow_prec()-1))}, mk_pi(*g_arrow_binder, Var(1), Var(1), binder_info()));
```
Comment (gebner) [inline src/frontends/lean/builtin_exprs.cpp:1097]: I wonder if we shouldn't change the `mk_arrow` function instead.

**Comment by gebner** [comment]
Replaced by #490.

**Pull request closed without merging**